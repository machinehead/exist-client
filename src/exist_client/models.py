from ._exist_io_client.models import Attribute, AttributeValue

__all__ = ["Attribute", "AttributeValue"]
