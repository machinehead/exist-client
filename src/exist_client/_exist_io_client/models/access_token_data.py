from typing import Any, Dict, List, Type, TypeVar

from attrs import define as _attrs_define
from attrs import field as _attrs_field

T = TypeVar("T", bound="AccessTokenData")


@_attrs_define
class AccessTokenData:
    """
    Attributes:
        grant_type (str):
        code (str):
        client_id (str):
        client_secret (str):
    """

    grant_type: str
    code: str
    client_id: str
    client_secret: str
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        grant_type = self.grant_type
        code = self.code
        client_id = self.client_id
        client_secret = self.client_secret

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update(
            {
                "grant_type": grant_type,
                "code": code,
                "client_id": client_id,
                "client_secret": client_secret,
            }
        )

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        grant_type = d.pop("grant_type")

        code = d.pop("code")

        client_id = d.pop("client_id")

        client_secret = d.pop("client_secret")

        access_token_data = cls(
            grant_type=grant_type,
            code=code,
            client_id=client_id,
            client_secret=client_secret,
        )

        access_token_data.additional_properties = d
        return access_token_data

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
