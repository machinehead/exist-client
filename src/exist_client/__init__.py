from .client import ExistClient

__all__ = ["ExistClient"]
