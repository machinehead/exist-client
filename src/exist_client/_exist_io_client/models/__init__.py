""" Contains all the data models used in inputs/outputs """

from .access_token_data import AccessTokenData
from .attribute import Attribute
from .attribute_by_name import AttributeByName
from .attribute_by_template import AttributeByTemplate
from .attribute_group import AttributeGroup
from .attribute_value import AttributeValue
from .attribute_values_get_result import AttributeValuesGetResult
from .attributes_acquire_result import AttributesAcquireResult
from .attributes_acquire_result_failed_item import AttributesAcquireResultFailedItem
from .attributes_acquire_result_success_item import AttributesAcquireResultSuccessItem
from .attributes_get_result import AttributesGetResult
from .attributes_update_result import AttributesUpdateResult
from .attributes_update_result_failed_item import AttributesUpdateResultFailedItem
from .date_value import DateValue
from .error_mixin import ErrorMixin
from .paginated_response import PaginatedResponse
from .tokens import Tokens
from .user_profile import UserProfile

__all__ = (
    "AccessTokenData",
    "Attribute",
    "AttributeByName",
    "AttributeByTemplate",
    "AttributeGroup",
    "AttributesAcquireResult",
    "AttributesAcquireResultFailedItem",
    "AttributesAcquireResultSuccessItem",
    "AttributesGetResult",
    "AttributesUpdateResult",
    "AttributesUpdateResultFailedItem",
    "AttributeValue",
    "AttributeValuesGetResult",
    "DateValue",
    "ErrorMixin",
    "PaginatedResponse",
    "Tokens",
    "UserProfile",
)
