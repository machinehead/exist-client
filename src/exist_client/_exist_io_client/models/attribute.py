from typing import TYPE_CHECKING, Any, Dict, List, Optional, Type, TypeVar

from attrs import define as _attrs_define
from attrs import field as _attrs_field

if TYPE_CHECKING:
    from ..models.attribute_group import AttributeGroup


T = TypeVar("T", bound="Attribute")


@_attrs_define
class Attribute:
    """
    Attributes:
        name (str):
        label (str):
        group (AttributeGroup):
        value_type_description (str):
        template (Optional[str]):
    """

    name: str
    label: str
    group: "AttributeGroup"
    value_type_description: str
    template: Optional[str]
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        name = self.name
        label = self.label
        group = self.group.to_dict()

        value_type_description = self.value_type_description
        template = self.template

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update(
            {
                "name": name,
                "label": label,
                "group": group,
                "value_type_description": value_type_description,
                "template": template,
            }
        )

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.attribute_group import AttributeGroup

        d = src_dict.copy()
        name = d.pop("name")

        label = d.pop("label")

        group = AttributeGroup.from_dict(d.pop("group"))

        value_type_description = d.pop("value_type_description")

        template = d.pop("template")

        attribute = cls(
            name=name,
            label=label,
            group=group,
            value_type_description=value_type_description,
            template=template,
        )

        attribute.additional_properties = d
        return attribute

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
