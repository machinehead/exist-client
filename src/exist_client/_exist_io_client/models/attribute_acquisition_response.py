from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar

from attrs import define as _attrs_define
from attrs import field as _attrs_field

if TYPE_CHECKING:
    from ..models.attribute_acquisition_response_failed_item import (
        AttributeAcquisitionResponseFailedItem,
    )
    from ..models.attribute_acquisition_response_success_item import (
        AttributeAcquisitionResponseSuccessItem,
    )


T = TypeVar("T", bound="AttributeAcquisitionResponse")


@_attrs_define
class AttributeAcquisitionResponse:
    """
    Attributes:
        success (List['AttributeAcquisitionResponseSuccessItem']):
        failed (List['AttributeAcquisitionResponseFailedItem']):
    """

    success: List["AttributeAcquisitionResponseSuccessItem"]
    failed: List["AttributeAcquisitionResponseFailedItem"]
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        success = []
        for success_item_data in self.success:
            success_item = success_item_data.to_dict()

            success.append(success_item)

        failed = []
        for failed_item_data in self.failed:
            failed_item = failed_item_data.to_dict()

            failed.append(failed_item)

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update(
            {
                "success": success,
                "failed": failed,
            }
        )

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.attribute_acquisition_response_failed_item import (
            AttributeAcquisitionResponseFailedItem,
        )
        from ..models.attribute_acquisition_response_success_item import (
            AttributeAcquisitionResponseSuccessItem,
        )

        d = src_dict.copy()
        success = []
        _success = d.pop("success")
        for success_item_data in _success:
            success_item = AttributeAcquisitionResponseSuccessItem.from_dict(
                success_item_data
            )

            success.append(success_item)

        failed = []
        _failed = d.pop("failed")
        for failed_item_data in _failed:
            failed_item = AttributeAcquisitionResponseFailedItem.from_dict(
                failed_item_data
            )

            failed.append(failed_item)

        attribute_acquisition_response = cls(
            success=success,
            failed=failed,
        )

        attribute_acquisition_response.additional_properties = d
        return attribute_acquisition_response

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
