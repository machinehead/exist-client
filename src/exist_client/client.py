import itertools
from typing import Any, Callable, Iterable, Optional, Protocol, TypeVar

from ._exist_io_client import AuthenticatedClient
from ._exist_io_client.api.default import (
    accounts_profile,
    attribute_values_get,
    attributes_acquire,
    attributes_get,
    attributes_update,
)

# TODO: move to .models
from ._exist_io_client.models import (
    AttributeByName,
    AttributeByTemplate,
    AttributesAcquireResult,
    AttributesUpdateResult,
    DateValue,
    UserProfile,
)
from .models import Attribute, AttributeValue

EXIST_IO_BASE_URL = "https://exist.io"

T = TypeVar("T")


class PaginatedResponse(Protocol[T]):
    results: list[T]
    next_: Optional[str]


PaginatedApi = Callable[..., Optional[PaginatedResponse[T]]]


class ExistClient:
    def __init__(self, *, token: str):
        self.client = AuthenticatedClient(
            EXIST_IO_BASE_URL,
            token,
            raise_on_unexpected_status=True,
            follow_redirects=True,
        )

    def get_profile(self) -> Optional[UserProfile]:
        return accounts_profile.sync(client=self.client)

    def _paginate(self, api: PaginatedApi[T], **kwargs: Any) -> Iterable[T]:
        for page in itertools.count(start=1):
            args = kwargs | dict(
                client=self.client,
                page=page,
            )
            resp = api(**args)
            assert resp is not None
            yield from resp.results
            if resp.next_ is None:
                break

    def get_attributes(
        self,
        *,
        attributes: Optional[list[str]] = None,
        groups: Optional[list[str]] = None,
        manual: Optional[bool] = None,
        owned: Optional[bool] = None,
    ) -> list[Attribute]:
        return list(
            self._paginate(
                attributes_get.sync,
                attributes=",".join(attributes) if attributes is not None else None,
                groups=",".join(groups) if groups is not None else None,
                manual=manual,
                owned=owned,
            )
        )

    def get_attribute_values(self, *, attribute: str) -> list[DateValue]:
        return list(self._paginate(attribute_values_get.sync, attribute=attribute))

    def acquire_attributes(
        self,
        *,
        acquisitions: list[AttributeByName | AttributeByTemplate],
    ) -> Optional[AttributesAcquireResult]:
        return attributes_acquire.sync(client=self.client, json_body=acquisitions)

    def update_attributes(
        self, *, updates: list[AttributeValue]
    ) -> Optional[AttributesUpdateResult]:
        return attributes_update.sync(client=self.client, json_body=updates)
