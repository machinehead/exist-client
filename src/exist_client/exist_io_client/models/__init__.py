""" Contains all the data models used in inputs/outputs """

from .access_token_data import AccessTokenData
from .attribute_acquisition import AttributeAcquisition
from .attribute_acquisition_response import AttributeAcquisitionResponse
from .attribute_acquisition_response_failed_item import (
    AttributeAcquisitionResponseFailedItem,
)
from .attribute_acquisition_response_success_item import (
    AttributeAcquisitionResponseSuccessItem,
)
from .attribute_update import AttributeUpdate
from .attributes_update_response import AttributesUpdateResponse
from .attributes_update_response_failed_item import AttributesUpdateResponseFailedItem
from .attributes_update_response_success_item import AttributesUpdateResponseSuccessItem
from .get_attributes_response_200 import GetAttributesResponse200
from .get_attributes_response_200_results_item import (
    GetAttributesResponse200ResultsItem,
)
from .get_attributes_response_200_results_item_group import (
    GetAttributesResponse200ResultsItemGroup,
)
from .tokens import Tokens
from .user_profile import UserProfile

__all__ = (
    "AccessTokenData",
    "AttributeAcquisition",
    "AttributeAcquisitionResponse",
    "AttributeAcquisitionResponseFailedItem",
    "AttributeAcquisitionResponseSuccessItem",
    "AttributesUpdateResponse",
    "AttributesUpdateResponseFailedItem",
    "AttributesUpdateResponseSuccessItem",
    "AttributeUpdate",
    "GetAttributesResponse200",
    "GetAttributesResponse200ResultsItem",
    "GetAttributesResponse200ResultsItemGroup",
    "Tokens",
    "UserProfile",
)
