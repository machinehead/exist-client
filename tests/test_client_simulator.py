from contextlib import nullcontext as does_not_raise
from typing import Any, Generator, Optional, TypeVar, Union

import pytest
import respx
from _pytest.python_api import RaisesContext
from flask import Flask, request
from flask_httpauth import HTTPTokenAuth
from respx import MockRouter, WSGIHandler
from syrupy.assertion import SnapshotAssertion

from exist_client import ExistClient
from exist_client._exist_io_client.errors import UnexpectedStatus
from exist_client._exist_io_client.models import (
    Attribute,
    AttributeGroup,
    AttributesGetResult,
    AttributesUpdateResult,
    AttributesUpdateResultFailedItem,
    AttributeValue,
)
from exist_client.client import EXIST_IO_BASE_URL

INVALID_TOKEN_RESPONSE_CONTENT = b'{"detail": "Invalid token"}'

_T = TypeVar("_T")

# TODO: support status 401
TOKEN_EXPECTATION_PARAMS = (
    "token,expectation",
    [
        ("known-token", does_not_raise()),
        (
            "unknown-token",
            pytest.raises(
                UnexpectedStatus,
                match=f"Unexpected status code: 401, content: {INVALID_TOKEN_RESPONSE_CONTENT!r}",
            ),
        ),
    ],
)


# https://docs.pytest.org/en/7.1.x/example/parametrize.html#parametrizing-conditional-raising
@pytest.mark.parametrize(*TOKEN_EXPECTATION_PARAMS)
def test_get_attributes(
    token: str,
    expectation: Union[does_not_raise[_T] | RaisesContext[Exception]],
    exist_api_simulator: MockRouter,
    snapshot: SnapshotAssertion,
) -> None:
    client = ExistClient(token=token)
    with expectation:
        assert client.get_attributes() == snapshot


@pytest.mark.parametrize(*TOKEN_EXPECTATION_PARAMS)
def test_update_attributes(
    token: str,
    expectation: Union[does_not_raise[_T] | RaisesContext[Exception]],
    exist_api_simulator: MockRouter,
    snapshot: SnapshotAssertion,
) -> None:
    client = ExistClient(token=token)
    with expectation:
        assert (
            client.update_attributes(
                updates=[
                    AttributeValue(name="myair_score", date="2023-12-29", value=57),
                    AttributeValue(name="nonexistent", date="2023-12-29", value=480),
                ]
            )
            == snapshot
        )


@pytest.fixture
def exist_api_simulator() -> Generator[MockRouter, None, None]:
    app = Flask("foobar")
    auth = HTTPTokenAuth(scheme="Bearer")

    tokens = {"known-token": "accepted"}

    @auth.verify_token
    def verify_token(token: str) -> Optional[str]:
        return tokens.get(token)

    @auth.error_handler
    def auth_error(status: int) -> tuple[bytes, int]:
        return INVALID_TOKEN_RESPONSE_CONTENT, status

    # {"template":"sleep","name":"sleep","label":"Time asleep","group":{"name":"sleep","label":"Sleep","priority":3},
    # "service":{"name":"garmin","label":"Garmin"},"active":true,"priority":1,"manual":false,"value_type":3,
    # "value_type_description":"Period (min)","available_services":[{"name":"garmin","label":"Garmin"}]}

    sleep_group = AttributeGroup(
        name="sleep",
        label="Sleep",
    )

    attributes: list[Attribute] = [
        Attribute(
            name="sleep",
            label="Time asleep",
            group=sleep_group,
            value_type_description="Period (min)",
            template="sleep",
        ),
        Attribute(
            name="myair_score",
            label="myAir score",
            group=sleep_group,
            value_type_description="Percentage",
            template=None,
        ),
    ]

    attribute_owned = {attr.name: False for attr in attributes}

    @app.get("/api/2/attributes/")
    @auth.login_required
    def get_attributes() -> dict[str, Any]:
        return AttributesGetResult(
            count=len(attributes),
            # TODO: filter by groups the client has access to
            results=attributes,
            # TODO: add pagination
            next_=None,
            previous=None,
        ).to_dict()

    @app.post("/api/2/attributes/update/")
    @auth.login_required
    def update_attributes() -> dict[str, Any]:
        request_json = request.json
        assert request_json is not None
        success = []
        failed = []
        for item in request_json:
            value = AttributeValue.from_dict(item)
            owned = attribute_owned.get(value.name)
            if owned is None:
                failed.append(
                    AttributesUpdateResultFailedItem(
                        error=f"User doesn't have an attribute named '{value.name}'",
                        error_code="not_found",
                        **value.to_dict(),
                    )
                )
            elif not owned:
                failed.append(
                    AttributesUpdateResultFailedItem(
                        error=f"Attribute '{value.name}' does not belong to this service",
                        error_code="unauthorised",
                        **value.to_dict(),
                    )
                )
            else:
                success.append(value)
        return AttributesUpdateResult(success=success, failed=failed).to_dict()

    with respx.mock(base_url=EXIST_IO_BASE_URL) as respx_mock:
        respx_mock.route().mock(side_effect=WSGIHandler(app))
        yield respx_mock
