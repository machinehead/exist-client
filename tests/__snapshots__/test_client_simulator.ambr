# serializer version: 1
# name: test_get_attributes[known-token-expectation0]
  list([
    Attribute(name='sleep', label='Time asleep', group=AttributeGroup(name='sleep', label='Sleep', additional_properties={}), value_type_description='Period (min)', template='sleep', additional_properties={}),
    Attribute(name='myair_score', label='myAir score', group=AttributeGroup(name='sleep', label='Sleep', additional_properties={}), value_type_description='Percentage', template=None, additional_properties={}),
  ])
# ---
# name: test_update_attributes[known-token-expectation0]
  AttributesUpdateResult(success=[], failed=[AttributesUpdateResultFailedItem(error="Attribute 'myair_score' does not belong to this service", error_code='unauthorised', date='2023-12-29', value=57, name='myair_score', additional_properties={}), AttributesUpdateResultFailedItem(error="User doesn't have an attribute named 'nonexistent'", error_code='not_found', date='2023-12-29', value=480, name='nonexistent', additional_properties={})], additional_properties={})
# ---
